src/mame/drivers/cgc7900.cpp

```diff
@@ -128,40 +170,99 @@ WRITE16_MEMBER( cgc7900_state::keyboard_w )
 {
 	/*
 
-	    bit     description
-
-	     0      real time clock
-	     1      RS-449 Tx ready
-	     2      BINT 2
-	     3      RS-232 Tx ready
-	     4      disk
-	     5      BINT 3
-	     6      bezel keys
-	     7      keyboard
-	     8      RS-449 Rx ready
-	     9      light pen
-	    10      BINT 4
-	    11      joystick
-	    12      vertical retrace
-	    13      BINT 5
-	    14      BINT 1
-	    15      RS-232 Rx ready
-
+	    bit     description		vec	level
+
+	     0      real time clock	4b	5
+	     1      RS-449 Tx ready	44	4
+	     2      BINT 2			4c	5
+	     3      RS-232 Tx ready	43	4
+	     4      disk			42	4
+	     5      BINT 3			4d	5
+	     6      bezel keys		45	4
+	     7      keyboard		4a	5
+	     8      RS-449 Rx ready	49	5
+	     9      light pen		46	4
+	    10      BINT 4			4e	5
+	    11      joystick		41	4
+	    12      vert. retrace	40	4
+	    13      BINT 5			4f	5
+	    14      BINT 1			47	4
+	    15      RS-232 Rx ready	48	5
+
+		default mask is 0x7e3f -- RS232 Rx, RS449 Rx, keyboard, bezel.
 	*/
 
+	if (m_int_mask != data)
+	{
+		u16 changed = m_int_mask ^ data;
+
+		LOG("i_mask: changed %04X -> %04X\n", m_int_mask, data);
+
+		for (int i = 0; i < 16; i++)
+		{
+			if (BIT(changed, i) && BIT(data, i))
+			{
+				LOG("i_mask: pin %d disabled, clearing irq\n", i);
+				irq_encoder(i, CLEAR_LINE);
+			}
+			if (BIT(changed, i) && !BIT(data, i))
+			{
+				LOG("i_mask: pin %d enabled, %s irq\n", i, BIT(m_int_active, i) ? "setting" : "clearing");
+				irq_encoder(i, BIT(m_int_active, i));
+			}
+		}
+	}
+
 	m_int_mask = data;
 }
 
+void cgc7900_state::irq_encoder(int pin, int state)
+{
+	if(state == ASSERT_LINE)
+		m_int_active |= (1 << pin);
+	else
+		m_int_active &= ~(1 << pin);
+
+	if (!BIT(m_int_mask, pin))
+	{
+		m_maincpu->set_input_line_and_vector(int_levels[pin], state, int_vectors[pin]);
+	}
+}
+
+WRITE_LINE_MEMBER(cgc7900_state::irq_0) { irq_encoder(0, state); }
+WRITE_LINE_MEMBER(cgc7900_state::irq_1) { irq_encoder(1, state); }
+WRITE_LINE_MEMBER(cgc7900_state::irq_2) { irq_encoder(2, state); }
+WRITE_LINE_MEMBER(cgc7900_state::irq_3) { irq_encoder(3, state); }
+WRITE_LINE_MEMBER(cgc7900_state::irq_4) { irq_encoder(4, state); }
+WRITE_LINE_MEMBER(cgc7900_state::irq_5) { irq_encoder(5, state); }
+WRITE_LINE_MEMBER(cgc7900_state::irq_6) { irq_encoder(6, state); }
+WRITE_LINE_MEMBER(cgc7900_state::irq_7) { irq_encoder(7, state); }
+WRITE_LINE_MEMBER(cgc7900_state::irq_8) { irq_encoder(8, state); }
+WRITE_LINE_MEMBER(cgc7900_state::irq_9) { irq_encoder(9, state); }
+WRITE_LINE_MEMBER(cgc7900_state::irq_a) { irq_encoder(0xa, state); }
+WRITE_LINE_MEMBER(cgc7900_state::irq_b) { irq_encoder(0xb, state); }
+WRITE_LINE_MEMBER(cgc7900_state::irq_c) { irq_encoder(0xc, state); }
+WRITE_LINE_MEMBER(cgc7900_state::irq_d) { irq_encoder(0xd, state); }
+WRITE_LINE_MEMBER(cgc7900_state::irq_e) { irq_encoder(0xe, state); }
+WRITE_LINE_MEMBER(cgc7900_state::irq_f) { irq_encoder(0xf, state); }
```

Comment:
Please use a template for this to reduce copy-paste, i.e. in the header at declaration point:
```
template <unsigned N> DECLARE_WRITE_LINE_MEMBER(irq) { irq_encoder(N, state); }
```

Then you can get each of them as `cgc7900_state::irq<0x0>`, `cgc7900_state::irq<0x1>`, ..., `cgc7900_state::irq<0xf>` and not need 16 of the things here.